include/seqan3/io/alignment_file/format_sam.hpp

```diff
@@ -0,0 +1,576 @@
+
+        if (!ranges::empty(get<1>(align)))
+        {
+            // compute possible distance from alignment end to sequence end
+            // which indicates soft clipping at the end.
+            // This should be replace by a free count_gaps function for
+            // aligned sequences which is more efficient if possible.
+            size_t off_end{seq.size() - offset};
+            for (auto chr : get<1>(align))
+                if (chr == gap::GAP)
+                    ++off_end;
+            off_end -= (get<1>(align)).size();
+
+            write_field(stream_it,
+                        detail::get_cigar_string(std::forward<align_type>(align),
+                                                 std::forward<offset_type>(offset),
+                                                 off_end),
+                        separator);
+        }
+        else
+        {
+            stream << "*\t";
+        }
+
+        write_field(stream_it, get<0>(std::forward<mate_type>(mate)), separator);
+
+        stream << get<1>(std::forward<mate_type>(mate)) << separator;
+
+        stream << get<2>(std::forward<mate_type>(mate)) << separator;
+
+        write_field(stream_it, std::forward<seq_type>(seq), separator);
+
+        write_range(stream_it, std::forward<qual_type>(qual));
+
+        write_tag_fields(stream, std::forward<tag_dict_type>(tag_dict), separator);
+
+        detail::write_eol(stream_it, options.add_carriage_return);
+    }
+
+protected:
+    //!/brief The format version string.
+    static constexpr char format_version[4] = "1.6";
+
+    //!\brief A variable that tracks whether the header_ptr as been written or not.
+    bool written_header{false};
+
+    /*!\brief Writes a field value to the stream adding a separator character.
+     * \tparam stream_it_t The stream iterator type.
+     * \tparam field_type  The type of the field value. Must model std::ranges::ForwardRange.
+     *
+     * \param[in,out] stream_it   The stream iterator to print to.
+     * \param[in]     field_value The value to print.
+     * \param[in]     separator   The field separator to append.
+     */
+    template <typename stream_it_t, typename field_type>
+        requires std::ranges::ForwardRange<field_type>
+    void write_field(stream_it_t & stream_it, field_type && field_value, char const separator)
+    {
+        write_range(stream_it, std::forward<field_type>(field_value));
+        stream_it = separator;
+    }
+
+
+    /*!\brief Writes a field value to the stream.
+     * \tparam stream_it_t The stream iterator type.
+     * \tparam field_type  The type of the field value. Must model std::ranges::ForwardRange.
+     *
+     * \param[in,out] stream_it   The stream iterator to print to.
+     * \param[in]     field_value The value to print.
+     */
+    template <typename stream_it_t, typename field_type>
+        requires std::ranges::ForwardRange<field_type>
+    void write_range(stream_it_t & stream_it, field_type && field_value)
```

Comment:
now there are two different functions? write_range AND write_field?

I feel like you did not address my previous points, yet...

Reply:
I deleted the field function and now the separator is printed separately